### PR TITLE
HWS: Introduce HW steering dump tool

### PR DESCRIPTION
**Pull request by hamdanigbaria**
Initial commit of HWS dump tool.

Signed-off-by: Hamdan Igbaria <hamdani@nvidia.com>

### CODE REVIEW

File: hw_steering_src/dr_definer.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    "src_qp_gvmi": [[('loopback_syndrome', 8), ('l3_type', 2), ('l4_type_bwc', 2), ('first_vlan_qualifier', 2), ('reserved_at_e', 1), ('functional_lb', 1), ('source_gvmi', 16)], [('force_lb', 1), ('ip_fragmented', 1), ('source_is_requestor', 1), ('reserved_at_23', 5), ('source_qp', 24)]],
+    "voq": [[('reserved_at_0', 24), ('ecn_ok', 1), ('congestion', 1), ('profile', 2), ('internal_prio', 4)]],
+    "NOT_DEFINED": [],
+    }
```
Comment (muhamadSammar) [inline hw_steering_src/dr_definer.py:22]: lets have new file "definer_parsers" and use for each definer a parser function. like:
mlx5_definer_parse_ib_l4
mlx5_definer_parse_ipv4_src_dst
...

File: hw_steering_src/dr_definer.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+dr_hl_array = [
```
Comment (muhamadSammar) [inline hw_steering_src/dr_definer.py:24]: call definer pasrer function defined above

File: hw_steering_src/dr_definer.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    ("ipv6_addr", 3),
+    ("ipv6_addr", 0),
+    ("ipv6_addr", 1),
+    ("ipv6_addr", 2),
+    ("ipv6_addr", 3),
+    ("ipv6_addr", 0),
+    ("ipv6_addr", 1),
+    ("ipv6_addr", 2),
+    ("ipv6_addr", 3),
+    ]
```
Comment (muhamadSammar) [inline hw_steering_src/dr_definer.py:109]: do we cover "unsupported" definer ?

**Review by muhamadSammar — CHANGES_REQUESTED** [review]

File: hw_steering_src/dr_hl.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    return dr_hl_dw_mask_parser(dw_fields[offset], mask)
+
+def dr_hl_eth_l2_parser(hl_index, mask):
+    offset = hl_index % 4
+    dw_fields = [
+                 [('dmac_47_16', 32)],
+                 [('dmac_15_0', 16), ('l3_ethertype', 16)],
+                 [('reserved_at_40', 1), ('sx_sniffer', 1), ('functional_lb', 1), ('ip_fragmented', 1), ('qp_type', 2), ('encap_type', 2), ('port_number', 2), ('l3_type', 2), ('l4_type_bwc', 2), ('first_vlan_qualifier', 2), ('first_priority', 3), ('first_cfi', 1), ('first_vlan_id', 12)],
+                 [('l4_type', 4), ('reserved_at_64', 2), ('ipsec_layer', 2), ('l2_type', 2), ('force_lb', 1), ('l2_ok', 1), ('l3_ok', 1), ('l4_ok', 1), ('second_vlan_qualifier', 2), ('second_priority', 3), ('second_cfi', 1), ('second_vlan_id', 12)]
+                ]
```
[the remark this answers is not in the document]
Comment (muhamadSammar) [reply hw_steering_src/dr_hl.py:53]: I think it's easier to debug with pdp this way, and it looks more like the excel file.
can we use same offset as in the excel file? and for inner take index - dic.keys.num (4 in this case) ?

File: mlx_hw_steering_parser.py [hunk 5]
```diff
@@ -118,9 +119,10 @@ def parse_args():
     if (args.dump_hw_resources):
         print("-hw is not supported yet.")
         sys.exit(0)
-    if (args.dpdk_pid != -1):
-        print("-pid is not supported yet.")
-        sys.exit(0)
+    if (args.dpdk_pid > 0):
+        if dr_trigger.trigger_dump(args.dpdk_pid, args.dpdk_port, args.FILEPATH, 0) is None:
```
Comment (muhamadSammar) [inline mlx_hw_steering_parser.py:123]: lets check after trigger, when we open the dump file, this way we will cover both cases, triggered and non-triggered

**Review by muhamadSammar — APPROVED** [review]

**Review by muhamadSammar — APPROVED** [review]

**Review by muhamadSammar — APPROVED** [review]

**Review by muhamadSammar — APPROVED** [review]

**Review by muhamadSammar — CHANGES_REQUESTED** [review]
lets check if we got an empty file on non-triggered mode

**Review by muhamadSammar — APPROVED** [review]

File: hw_steering_src/dr_definer.py [hunk 6]
```diff
@@ -95,11 +85,7 @@ def definer_dws_parser(self):
         for i in range(6):
```
Comment (muhamadSammar) [inline hw_steering_src/dr_definer.py:85]: I think we need to fix all magic numbers, lets define them above or in separate file

Comment (muhamadSammar) [reply hw_steering_src/dr_definer.py:85]: range(6), 6 is the number of dws in a definer?

File: hw_steering_src/dr_definer.py [hunk 7]
```diff
@@ -113,15 +99,31 @@ def definer_bytes_parser(self):
             hl_byte_index = int(self.data[byte_selector], 16)
             hl_dw_index = hl_byte_index / 4
             hl_byte_offset = hl_byte_index % 4
-            mask = ((3 - hl_byte_offset) * 8) * "0"
-            mask += hex_to_bin_str(self.data["mask_tag_" + str(i)])
-            mask += (hl_byte_offset * 8) * "0"
+            mask = (hl_byte_offset * 8) * "0"
+            mask += hex_to_bin_str(self.data["byte_mask_tag_" + str(i)], 8)
+            mask += ((3 - hl_byte_offset) * 8) * "0"
```
Comment (muhamadSammar) [inline hw_steering_src/dr_definer.py:104]: 8 - BYTE_SZ
3?

**Review by muhamadSammar — CHANGES_REQUESTED** [review]

**Review by muhamadSammar — APPROVED** [review]

File: hw_steering_src/dr_definer.py (hunk 7 above)
Comment (hamdanigbaria) [reply hw_steering_src/dr_definer.py:104]: done

File: hw_steering_src/dr_definer.py (hunk 6 above)
Comment (hamdanigbaria) [reply hw_steering_src/dr_definer.py:85]: Done

File: hw_steering_src/dr_hl.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hamdanigbaria) [reply hw_steering_src/dr_hl.py]: Done

File: hw_steering_src/dr_hl.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    return fields_arr
+
+#The following functions decribes the headers layout structs
+#The functions recieve hl_index which is the DWs offset according to the headers layout and the DW mask
+#The offset parameter inside the functions is the DW offset inside the specific headers layout struct
+#oofset = (hl_index - "the starting offset of the hl struct") % ("hl struct size")
+#The dw_fields describes the specific hl struct, such that each array element in the array describes a DW and each DW described as (field name, bits size)
+#Returns the masked fields as an array (by calling dr_hl_dw_mask_parser function)
+
+def dr_hl_ib_l4_parser(hl_index, mask):
```
[the remark this answers is not in the document]
Comment (hamdanigbaria) [reply hw_steering_src/dr_hl.py]: Done

File: hw_steering_src/dr_definer.py (hunk 3 above)
Comment (hamdanigbaria) [reply hw_steering_src/dr_definer.py]: yes

File: hw_steering_src/dr_definer.py (hunk 1 above)
Comment (hamdanigbaria) [reply hw_steering_src/dr_definer.py]: done

File: hw_steering_src/dr_definer.py (hunk 2 above)
Comment (hamdanigbaria) [reply hw_steering_src/dr_definer.py]: done

**Pull request closed without merging**